torch/utils/data/dataloader.py

```diff
@@ -69,7 +69,8 @@ def _worker_manager_loop(in_queue, out_queue, done_event, pin_memory):
             continue
         idx, batch = r
         try:
-            if pin_memory:
+            if pin_memory and torch.cuda.is_available():
+                torch.cuda.set_device(device_id)
```

Comment:
TIL that pinned memory is GPU-specific

Reply:
Whaaaat? 😮

Reply:
The joys of default context creation :-) Pinned memory is not GPU specific (as in, if it is pinned it can be asynchronously transferred to any GPU), but a call to cudaHostAlloc will initialize context on the default device. So, if one runs distributed on a single node, with multi-process dataloader, even though one would set appropriate device in the main process, dataloader processes would initialize context on device 0.

Reply:
That makes sense, thanks Natalia. Can we move this code outside of the main loop? The device id can’t change during the run time, so it should be sufficient to set it once